paasta_tools/autoscaling/autoscaling_service_lib.py

```diff
@@ -374,17 +374,24 @@ def mesos_cpu_metrics_provider(
     monkey.patch_socket()
     jobs = [gevent.spawn(task.stats_callable) for task in mesos_tasks]
     gevent.joinall(jobs, timeout=60)
-    mesos_tasks = dict(zip([task['id'] for task in mesos_tasks], [job.value for job in jobs]))
+    mesos_tasks_first_run = dict(zip([task['id'] for task in mesos_tasks], [job.value for job in jobs]))
 
     current_time = int(datetime.now().strftime('%s'))
     time_delta = current_time - last_time
 
+    # Mesos slave statistics endpoint returns a crazy CPU value when a container is
+    # in the process of being killed. So we fetch twice, and use the lower of the two.
+    jobs = [gevent.spawn(task.stats_callable) for task in mesos_tasks]
+    gevent.joinall(jobs, timeout=60)
+    mesos_tasks_second_run = dict(zip([task['id'] for task in mesos_tasks], [job.value for job in jobs]))
+
     mesos_cpu_data = {}
-    for task_id, stats in mesos_tasks.items():
-        if stats is not None:
+    for task_id, stats in mesos_tasks_first_run.items():
+        stats2 = mesos_tasks_second_run[task_id]
```

Comment:
Isn't there a potential issue where the tasks change between the first and second run causing this dictionary lookup to fail? Maybe use `mesos_tasks_second_run.get('task_id')` instead.

Reply:
good catch